pkg/scheduler/backend/api_dispatcher/call_controller.go

```diff
@@ -0,0 +1,300 @@
+// is currently in-flight, the call is skipped and an ErrCallSkipped is returned.
+//
+// If the call is not skipped, it is stored internally. If no previous call for the object
+// was being tracked (oldCallPresent is false), the object's UID is added to the processing
+// queue to be picked up by a worker.
+//
+// This logic ensures only one call per object is active in the queue at a time.
+// Must be called under cc.lock.
+func (cc *callController) enqueue(apiCall *queuedAPICall, oldCallPresent bool) error {
+	noOp := apiCall.IsNoOp()
+	if noOp && !cc.inFlightEntities.Has(apiCall.UID()) {
+		// The call can be skipped, as it is a no-op and the old call is not in-flight.
+		if oldCallPresent {
+			cc.removePending(apiCall.UID())
+		}
+		apiCall.sendOnFinish(fmt.Errorf("call does not need to be executed because it has no effect: %w", fwk.ErrCallSkipped))
+		return fmt.Errorf("call does not need to be executed because it has no effect: %w", fwk.ErrCallSkipped)
+	}
+
+	objectUID := apiCall.UID()
+	cc.apiCalls[objectUID] = apiCall
+	if oldCallPresent {
+		// If another API call for this object is already present (i.e., is pending or in-flight),
+		// don't add this new call to the queue. The new call will be processed
+		// after the currently in-flight call is finalized.
+		return nil
+	}
+	cc.callsQueue.WriteOne(objectUID)
+	cc.cond.Broadcast()
+	return nil
+}
+
+// removeFromQueue removes the objectUID from the queue and returns the recreated queue.
+func removeFromQueue(queue *buffer.Ring[types.UID], objectUID types.UID) *buffer.Ring[types.UID] {
+	newQueue := buffer.NewRing[types.UID](buffer.RingOptions{
+		InitialSize: queue.Len(),
+		NormalSize:  queue.Cap(),
+	})
+	for {
+		uid, ok := queue.ReadOne()
+		if !ok {
+			break
+		}
+		if uid != objectUID {
+			newQueue.WriteOne(uid)
+		}
+	}
+	return newQueue
+}
+
+// removePending removes a pending API call for the given objectUID from the queue and its associated data.
+// This function is intended to be used on calls that have not yet been popped for execution
+// (i.e., are not in-flight).
+// Must be called under cc.lock.
+func (cc *callController) removePending(objectUID types.UID) {
+	delete(cc.apiCalls, objectUID)
+	if !cc.inFlightEntities.Has(objectUID) {
+		cc.callsQueue = *removeFromQueue(&cc.callsQueue, objectUID)
+	}
+}
+
+// add adds a new apiCall to the queue.
+// If an API call for the same object is already present in the queue,
+// it tries to skip, overwrite, or merge the calls based on their relevance and type.
+func (cc *callController) add(apiCall *queuedAPICall) error {
+	cc.lock.Lock()
+	defer cc.lock.Unlock()
+
+	apiCall.callID = cc.callIDCounter
+	cc.callIDCounter++
+
+	oldAPICall, ok := cc.apiCalls[apiCall.UID()]
+	if ok {
+		err := cc.reconcile(oldAPICall, apiCall)
+		if err != nil {
+			return err
+		}
+	}
+	return cc.enqueue(apiCall, ok)
+}
+
+// pop pops the first object UID from the queue and returns the corresponding API call details.
```

Comment:
nit: 

```suggestion
// pop pops the first object UID from the queue and returns the corresponding API call details.
// Note: The caller must call finalize() on the poped apiCall, otherwise we would result in the memory leak. 
```